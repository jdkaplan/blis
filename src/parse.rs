use std::path::Path;

pub type ParseResult<T> = Result<T, (Option<T>, ParseErrors)>;

pub type ParseErrors = Vec<ParseError>;

#[derive(thiserror::Error, Debug)]
#[error("parse error")]
pub struct ParseError;

pub struct File {
    pub name: String,
    pub decls: Vec<Decl>,
}

pub enum Decl {
    Stmt(Stmt),
}

pub enum Stmt {
    Expr(Expr),
}

pub enum Expr {
    Literal(Literal),
}

pub enum Literal {
    U64(u64),
}

pub struct Parser;

impl Parser {
    pub fn parse(path: &Path, source: &str) -> ParseResult<File> {
        let mut errors = Vec::new();
        let mut decls = Vec::new();

        if source.trim() == "1" {
            let decl = Decl::Stmt(Stmt::Expr(Expr::Literal(Literal::U64(1))));
            decls.push(decl);
        } else {
            errors.push(ParseError);
        }

        let file = File {
            name: path.to_string_lossy().into(),
            decls,
        };

        if errors.is_empty() {
            Ok(file)
        } else {
            Err((Some(file), errors))
        }
    }
}
