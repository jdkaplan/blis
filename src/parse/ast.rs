pub struct Program {
    pub decls: Vec<Declaration>,
}

pub enum Declaration {
    Let(Let),

    Statement(Statement),
}

pub enum Statement {
    Expression(Expression),
}

pub enum Expression {
    Block(Block),
    If(If),

    // TODO: Atom?
    Identifier(Identifier),
    Literal(Literal),
}

impl Expression {
    pub fn self_terminating(&self) -> bool {
        match self {
            Expression::Block(_) | Expression::If(_) => true,
            Expression::Identifier(_) | Expression::Literal(_) => false,
        }
    }
}

pub struct Let {
    pub ident: Identifier,
    pub expr: Expression,
}

pub struct Block {
    pub decls: Vec<Declaration>,
    pub expr: Option<Box<Expression>>,
}

pub struct If {
    pub condition: Box<Expression>,
    pub consequent: Block,
    pub alternative: Option<Block>,
}

#[derive(Debug, Clone)]
pub struct Identifier {
    pub name: String,
}

impl Identifier {
    pub fn new(name: impl Into<String>) -> Self {
        Self { name: name.into() }
    }
}

pub enum Literal {
    Nil,
    Boolean(bool),
    Integer(u64),
    Float(f64),
    String(String),
}
