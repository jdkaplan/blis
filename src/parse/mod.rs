pub mod ast;
pub mod lexer;
pub mod parser;
pub mod token;

pub use lexer::Lexer;
pub use parser::{ParseError, ParseErrors, Parser};
pub use token::{Lexeme, LexemeOwned, Token};
