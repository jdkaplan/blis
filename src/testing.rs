use std::collections::HashMap;
use std::path::{Path, PathBuf};

use once_cell::sync::Lazy;
use regex::Regex;

/// Return the path relative to the Cargo manifest.
pub(crate) fn relative_path(path: &Path) -> String {
    let crate_root = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    let relative = path.strip_prefix(crate_root).unwrap();
    relative.to_string_lossy().to_string()
}

pub(crate) fn snapshot_name(path: &Path, name: &str) -> String {
    format!("{}.{}", relative_path(path), name)
}

pub(crate) fn replace_object_ids(original: String) -> String {
    let mut ids = HashMap::new();
    let mut counter = 0u64;
    let mut out = String::with_capacity(original.len());

    static RE: Lazy<Regex> =
        Lazy::new(|| Regex::new(r"<(?<ty>.*?) (?<id>0x[0-9a-fA-F]+)>").unwrap());

    let mut last = 0;
    for cap in RE.captures_iter(&original) {
        // Add in anything between the previous match and this one.
        let matched = cap.get(0).expect("full match");
        out.push_str(&original[last..matched.start()]);
        last = matched.end();

        // Add in the replacement from this match.
        let ty = cap.name("ty").expect("named group").as_str();
        let old = cap.name("id").expect("named group").as_str();
        let new = ids.entry(old).or_insert_with(|| {
            counter += 1;
            counter
        });
        out.push_str(&format!("<{} 0x{:012x}>", ty, new));
    }

    // Keep anything after the last match.
    if last < original.len() {
        out.push_str(&original[last..]);
    }

    out
}
