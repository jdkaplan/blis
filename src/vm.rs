use crate::bytecode::{Chunk, Constant, Op, OpError};

#[derive(Debug)]
pub enum Value {
    Nil,
    Boolean(bool),
    Integer(u64),
    Float(f64),
    String(String),
}

impl From<Constant> for Value {
    fn from(constant: Constant) -> Self {
        match constant {
            Constant::Integer(v) => Value::Integer(v),
            Constant::Float(v) => Value::Float(v),
            Constant::String(v) => Value::String(v),
        }
    }
}

#[derive(Default)]
pub struct Vm {
    stack: Vec<Value>,
}

pub type VmResult<T> = Result<T, VmError>;

#[derive(thiserror::Error, Debug)]
pub enum VmError {
    #[error(transparent)]
    Op(#[from] OpError),

    #[error(
        "stack error: expected value at depth {}, but stack length was {}",
        depth,
        len
    )]
    NoValue { depth: usize, len: usize },
}

impl Vm {
    pub fn new() -> Self {
        Self { stack: Vec::new() }
    }

    pub fn interpret(&mut self, chunk: Chunk) -> VmResult<()> {
        let mut pc = 0;

        loop {
            let op = Op::scan(&chunk.code[pc..])?;
            let Some(op) = op else {
                return Ok(());
            };
            pc += op.size_bytes();

            match op {
                Op::Return => {
                    dbg!(&self.stack);
                    return Ok(());
                }

                Op::Pop => {
                    dbg!(self.stack.pop());
                }
                Op::PopN(n) => {
                    let n = n as usize;
                    let len = self.stack.len();

                    let new_len = len.checked_sub(n);
                    let new_len = new_len.ok_or(VmError::NoValue { depth: n, len })?;

                    dbg!(&self.stack[new_len..]);
                    self.stack.truncate(new_len);
                }

                Op::Constant(idx) => {
                    let constant = &chunk.constants[idx as usize];
                    self.stack.push(Value::from(constant.clone()));
                }

                Op::Nil => {
                    self.stack.push(Value::Nil);
                }
                Op::False => {
                    self.stack.push(Value::Boolean(false));
                }
                Op::True => {
                    self.stack.push(Value::Boolean(true));
                }
            }
        }
    }
}
