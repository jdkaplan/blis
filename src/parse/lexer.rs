use tracing::instrument;

use crate::parse::{Lexeme, Token};

pub struct Lexer<'source> {
    source: &'source str,
    start: usize,
    current: usize,
    line: usize,
}

impl std::fmt::Debug for Lexer<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Lexer")
            .field("source", &format!("({} bytes)", self.source.bytes().len()))
            .field("start", &self.start)
            .field("current", &self.current)
            .field("line", &self.line)
            .finish()
    }
}

impl<'source> Iterator for Lexer<'source> {
    type Item = Lexeme<'source>;

    fn next(&mut self) -> Option<Self::Item> {
        Some(self.scan_next())
    }
}

impl<'source> Lexer<'source> {
    pub fn new(source: &'source str) -> Self {
        Self {
            source,
            start: 0,
            current: 0,
            line: 1,
        }
    }

    #[instrument(level = "trace", ret)]
    pub fn scan_next(&mut self) -> Lexeme<'source> {
        self.skip_ignored();

        self.start = self.current;

        if self.at_eof() {
            return self.make(Token::Eof);
        }

        match self.advance() {
            // One character, no lookahead
            '(' => self.make(Token::LeftParen),
            ')' => self.make(Token::RightParen),
            '[' => self.make(Token::LeftBracket),
            ']' => self.make(Token::RightBracket),
            '{' => self.make(Token::LeftBrace),
            '}' => self.make(Token::RightBrace),
            ';' => self.make(Token::Semicolon),
            ',' => self.make(Token::Comma),
            '.' => self.make(Token::Dot),
            '-' => self.make(Token::Minus),
            '+' => self.make(Token::Plus),
            '/' => self.make(Token::Slash),
            '*' => self.make(Token::Star),
            '%' => self.make(Token::Percent),

            // One or two characters, 1 lookahead
            '!' => {
                if self.take('=') {
                    self.make(Token::BangEqual)
                } else {
                    self.make(Token::Bang)
                }
            }
            '=' => {
                if self.take('=') {
                    self.make(Token::EqualEqual)
                } else {
                    self.make(Token::Equal)
                }
            }
            '<' => {
                if self.take('=') {
                    self.make(Token::LessEqual)
                } else {
                    self.make(Token::Less)
                }
            }
            '>' => {
                if self.take('=') {
                    self.make(Token::GreaterEqual)
                } else {
                    self.make(Token::Greater)
                }
            }

            '"' => self.scan_string(),

            // TODO: Allow non-ASCII identifiers
            c if is_alpha(c) => self.scan_identifier(),
            c if is_digit(c) => self.scan_number(),

            _ => self.error("unexpected character"),
        }
    }

    fn skip_ignored(&mut self) {
        loop {
            match self.peek() {
                // Whitespace
                ' ' | '\r' | '\t' => {
                    self.advance();
                }
                '\n' => {
                    self.line += 1;
                    self.advance();
                }

                // Comments
                '/' => {
                    if self.peek_next() == '/' {
                        while self.peek() != '\n' && !self.at_eof() {
                            self.advance();
                        }
                    } else {
                        return;
                    }
                }

                _ => return,
            }
        }
    }

    fn scan_identifier(&mut self) -> Lexeme<'source> {
        while is_alpha(self.peek()) || is_digit(self.peek()) {
            self.advance();
        }

        let text = &self.source[self.start..self.current];

        let token = keyword_or_identifier(text);
        self.make(token)
    }

    fn scan_number(&mut self) -> Lexeme<'source> {
        while is_digit(self.peek()) {
            self.advance();
        }

        let mut has_fraction = false;

        // Only consume the dot if there are also more digits afterward. Otherwise, it indicates a
        // method call rather than a decimal number.
        if self.peek() == '.' && is_digit(self.peek_next()) {
            self.advance(); // `.`
            has_fraction = true;

            while is_digit(self.peek()) {
                self.advance();
            }
        }

        if has_fraction {
            self.make(Token::Float)
        } else {
            self.make(Token::Integer)
        }
    }

    fn scan_string(&mut self) -> Lexeme<'source> {
        while self.peek() != '"' && !self.at_eof() {
            if self.peek() == '\n' {
                self.line += 1;
            }

            // There are two escape sequences that have to be handled here in the lexer
            // because they would otherwise cause strings to terminate early:
            //
            //  1. `\"` escapes the double-quote, allowing the string to continue.
            //  2. `\\` escapes the backslash, preventing `\\"` from escaping the quote.
            //
            // All other escapes are interpreted during parsing.
            if self.peek() == '\\' && matches!(self.peek_next(), '"' | '\\') {
                self.advance(); // peek \
                self.advance(); // peek_next
            } else {
                self.advance(); // peek
            }
        }

        if self.at_eof() {
            return self.error("unterminated string");
        }

        self.advance(); // peek `"`
        self.make(Token::String)
    }

    fn at_eof(&self) -> bool {
        self.peek() == '\0'
    }

    fn peek(&self) -> char {
        self.source[self.current..]
            .chars()
            .next()
            .unwrap_or_default()
    }

    fn peek_next(&self) -> char {
        if self.at_eof() {
            return '\0';
        }

        let mut chars = self.source[self.current..].chars();
        chars.next().unwrap(); // peek 0
        chars.next().unwrap_or_default() // peek 1
    }

    fn advance(&mut self) -> char {
        let c = self.peek();
        self.current += c.len_utf8();
        c
    }

    fn take(&mut self, expected: char) -> bool {
        if self.at_eof() {
            return false;
        }

        if self.peek() != expected {
            return false;
        }

        self.advance();
        true
    }

    fn make(&self, token: Token) -> Lexeme<'source> {
        let text = &self.source[self.start..self.current];

        Lexeme {
            token,
            text,
            line: self.line,
        }
    }

    fn error<'msg>(&self, msg: &'msg str) -> Lexeme<'msg> {
        Lexeme {
            token: Token::Error,
            text: msg,
            line: self.line,
        }
    }
}

fn is_alpha(c: char) -> bool {
    c.is_ascii_alphabetic() || c == '_'
}

fn is_digit(c: char) -> bool {
    c.is_ascii_digit()
}

fn keyword_or_identifier(text: &str) -> Token {
    match text {
        "and" => Token::And,
        "break" => Token::Break,
        "continue" => Token::Continue,
        "else" => Token::Else,
        "false" => Token::False,
        "func" => Token::Func,
        "if" => Token::If,
        "let" => Token::Let,
        "loop" => Token::Loop,
        "nil" => Token::Nil,
        "object" => Token::Object,
        "or" => Token::Or,
        "return" => Token::Return,
        "true" => Token::True,
        _ => Token::Identifier,
    }
}
