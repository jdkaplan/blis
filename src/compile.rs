use num_rational::BigRational;
use tracing::{debug, instrument};

use crate::bytecode::{Chunk, Constant, Op};
use crate::parse::ast::*;

#[derive(Debug)]
pub struct Compiler {
    chunk: Chunk,
    errors: Vec<CompileError>,

    locals: Locals,
}

#[derive(thiserror::Error, Debug)]
pub struct CompileErrors(Vec<CompileError>);

impl std::fmt::Display for CompileErrors {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        if let Some(err) = self.0.first() {
            write!(f, "{}", err)?;
        } else {
            return write!(f, "");
        }

        for err in &self.0[1..] {
            write!(f, "\n{}", err)?;
        }
        Ok(())
    }
}

#[derive(thiserror::Error, Debug)]
pub enum CompileError {
    #[error(
        "variable redeclared in the same scope: `{}`",
        .0.name
    )]
    DuplicateVariable(Identifier),

    #[error("{0}")]
    Other(String),
}

type CompileResult<T> = Result<T, CompileError>;

type Fallible<T> = Result<T, FailedCodegen>;

#[derive(Debug)]
struct FailedCodegen;

impl Compiler {
    fn new() -> Self {
        Self {
            chunk: Chunk::default(),
            errors: Vec::new(),
            locals: Locals::default(),
        }
    }

    fn error(&mut self, err: CompileError) -> FailedCodegen {
        debug!({ ?err }, "compile error");
        self.errors.push(err);
        FailedCodegen
    }

    #[instrument(level = "debug", ret)]
    pub fn compile(program: &Program) -> Result<Chunk, CompileErrors> {
        let mut compiler = Self::new();
        let _ = compiler.program(program); // Errors checked below

        if compiler.errors.is_empty() {
            for (idx, val) in compiler.chunk.constants.iter().enumerate() {
                debug!({ ?idx, ?val }, "constant");
            }

            for (idx, name) in compiler.chunk.globals.iter().enumerate() {
                debug!({ ?idx, ?name }, "global");
            }

            for res in compiler.chunk.iter_code() {
                let Ok((pc, op)) = res else {
                    debug!({ ?res }, "chunk error");
                    break;
                };
                debug!({ ?pc, ?op }, "code");
            }

            Ok(compiler.chunk)
        } else {
            Err(CompileErrors(compiler.errors))
        }
    }

    #[instrument(level = "trace")]
    fn program(&mut self, program: &Program) -> Fallible<()> {
        for decl in &program.decls {
            self.declaration(decl)?;
        }
        self.chunk.push(Op::Return);
        Ok(())
    }

    #[instrument(level = "trace")]
    fn block(&mut self, block: &Block) -> Fallible<()> {
        self.locals.begin_scope();

        for decl in &block.decls {
            self.declaration(decl)?;
        }

        if let Some(expr) = &block.expr {
            self.expression(expr)?;
        } else {
            self.chunk.push(Op::Nil);
        }

        self.locals.end_scope();
        Ok(())
    }

    #[instrument(level = "trace")]
    fn declaration(&mut self, decl: &Declaration) -> Fallible<()> {
        match decl {
            Declaration::Func(func) => self.decl_func(func),
            Declaration::Let(let_) => self.decl_let(let_),
            Declaration::Statement(stmt) => self.statement(stmt),
        }
    }

    #[instrument(level = "trace")]
    fn decl_func(&mut self, func: &Func) -> Fallible<()> {
        // - Declare the identifier in the current scope
        // - Mark it as initialized so it's usable within the func body
        // - Start a new scope containing all the param names
        // - Start a new chunk for the implementation.
        // - Compile the function into the chunk.
        // - Make a constant to hold the func header
        // - Load the constant and store it in the local/global name
        todo!("start here")
    }

    #[instrument(level = "trace")]
    fn decl_let(&mut self, let_: &Let) -> Fallible<()> {
        let slot = self
            .locals
            .reserve(let_.ident.clone())
            .map_err(|err| self.error(err))?;

        self.expression(&let_.expr)?;

        self.locals.mark_init(slot);
        Ok(())
    }

    #[instrument(level = "trace")]
    fn statement(&mut self, stmt: &Statement) -> Fallible<()> {
        match stmt {
            Statement::Assignment(assign) => self.stmt_assign(assign),

            Statement::Expression(expr) => {
                self.expression(expr)?;
                self.chunk.push(Op::Pop);
                Ok(())
            }
        }
    }

    #[instrument(level = "trace")]
    fn stmt_assign(&mut self, assign: &Assignment) -> Fallible<()> {
        let Place::Identifier(ident) = &assign.place;

        self.expression(&assign.expr)?;

        if let Some((slot, _local)) = self.locals.resolve(&ident.name) {
            self.chunk.push(Op::LocalSet(slot));
        } else {
            let global = self.chunk.make_global(ident.name.clone());
            self.chunk.push(Op::GlobalSet(global));
        }

        Ok(())
    }

    #[instrument(level = "trace")]
    fn expression(&mut self, expr: &Expression) -> Fallible<()> {
        let Expression::LogicOr(or) = expr;
        self.logic_or(or)
    }

    #[instrument(level = "trace")]
    fn logic_or(&mut self, or: &LogicOr) -> Fallible<()> {
        // This leaves the last value that was considered on top of the stack.

        // <a>
        self.logic_and(&or.first)?;

        let mut shorts = Vec::new();
        for and in &or.rest {
            // jump_true_peek 'short_circuit
            let short = self.chunk.prepare_jump(Op::JumpTruePeek(i16::MAX));
            shorts.push(short);

            // pop ; <a> was falsey
            self.chunk.push(Op::Pop);

            // <b>
            self.logic_and(and)?;
        }

        // 'short_circuit:
        for jump in shorts {
            self.chunk.set_jump_target(jump);
        }

        Ok(())
    }

    fn logic_and(&mut self, and: &LogicAnd) -> Fallible<()> {
        // This leaves the last value that was considered on top of the stack.

        // <a>
        self.equality(&and.first)?;

        let mut shorts = Vec::new();
        for eq in &and.rest {
            // jump_false_peek 'short_circuit
            let short = self.chunk.prepare_jump(Op::JumpFalsePeek(i16::MAX));
            shorts.push(short);

            // pop ; <a> was truthy
            self.chunk.push(Op::Pop);

            // <b>
            self.equality(eq)?;
        }

        // 'short_circuit:
        for jump in shorts {
            self.chunk.set_jump_target(jump);
        }

        Ok(())
    }

    fn equality(&mut self, eq: &Equality) -> Fallible<()> {
        match eq {
            Equality::Value(comp) => self.comparison(comp),

            Equality::Eq(a, b) => {
                self.equality(a)?;
                self.comparison(b)?;
                self.chunk.push(Op::Eq);
                Ok(())
            }

            Equality::Ne(a, b) => {
                self.equality(a)?;
                self.comparison(b)?;
                self.chunk.push(Op::Ne);
                Ok(())
            }
        }
    }

    fn comparison(&mut self, comp: &Comparison) -> Fallible<()> {
        match comp {
            Comparison::Value(term) => self.term(term),

            Comparison::Gt(a, b) => {
                self.comparison(a)?;
                self.term(b)?;
                self.chunk.push(Op::Gt);
                Ok(())
            }

            Comparison::Ge(a, b) => {
                self.comparison(a)?;
                self.term(b)?;
                self.chunk.push(Op::Ge);
                Ok(())
            }

            Comparison::Lt(a, b) => {
                self.comparison(a)?;
                self.term(b)?;
                self.chunk.push(Op::Lt);
                Ok(())
            }

            Comparison::Le(a, b) => {
                self.comparison(a)?;
                self.term(b)?;
                self.chunk.push(Op::Le);
                Ok(())
            }
        }
    }

    fn term(&mut self, term: &Term) -> Fallible<()> {
        match term {
            Term::Value(factor) => self.factor(factor),

            Term::Add(a, b) => {
                self.term(a)?;
                self.factor(b)?;
                self.chunk.push(Op::Add);
                Ok(())
            }

            Term::Sub(a, b) => {
                self.term(a)?;
                self.factor(b)?;
                self.chunk.push(Op::Sub);
                Ok(())
            }
        }
    }

    fn factor(&mut self, factor: &Factor) -> Fallible<()> {
        match factor {
            Factor::Value(unary) => self.unary(unary),

            Factor::Mul(a, b) => {
                self.factor(a)?;
                self.unary(b)?;
                self.chunk.push(Op::Mul);
                Ok(())
            }

            Factor::Div(a, b) => {
                self.factor(a)?;
                self.unary(b)?;
                self.chunk.push(Op::Div);
                Ok(())
            }

            Factor::Rem(a, b) => {
                self.factor(a)?;
                self.unary(b)?;
                self.chunk.push(Op::Rem);
                Ok(())
            }
        }
    }

    fn unary(&mut self, unary: &Unary) -> Fallible<()> {
        match unary {
            Unary::Value(call) => self.call(call),

            Unary::Not(a) => {
                self.unary(a)?;
                self.chunk.push(Op::Not);
                Ok(())
            }

            Unary::Neg(a) => {
                self.unary(a)?;
                self.chunk.push(Op::Neg);
                Ok(())
            }
        }
    }

    fn call(&mut self, call: &Call) -> Fallible<()> {
        match call {
            Call::Value(primary) => self.primary(primary),

            Call::Call(callee, args) => {
                self.call(callee)?;
                for arg in args {
                    self.expression(arg)?;
                }

                let arity: u8 = args.len().try_into().expect("at most 255 args");
                self.chunk.push(Op::Call(arity));
                Ok(())
            }

            Call::Index(obj, key) => {
                self.call(obj)?;
                self.expression(key)?;
                self.chunk.push(Op::Index);
                Ok(())
            }

            Call::Field(obj, field) => {
                self.call(obj)?;

                let name = Constant::String(field.name.clone());
                let id = self.chunk.add_constant(name);
                self.chunk.push(Op::Constant(id));
                Ok(())
            }
        }
    }

    #[instrument(level = "trace")]
    fn primary(&mut self, primary: &Primary) -> Fallible<()> {
        match primary {
            Primary::Block(block) => self.block(block),
            Primary::If(if_) => self.expr_if(if_),
            Primary::Atom(atom) => self.atom(atom),
        }
    }

    #[instrument(level = "trace")]
    fn atom(&mut self, atom: &Atom) -> Fallible<()> {
        match atom {
            Atom::Identifier(ident) => self.identifier(ident),
            Atom::Literal(lit) => self.literal(lit),
        }
    }

    #[instrument(level = "trace")]
    fn expr_if(&mut self, if_: &If) -> Fallible<()> {
        // <cond>
        self.expression(&if_.condition)?;

        // jump_false_pop 'skip_conseq [peek cond]
        let skip_conseq = self.chunk.prepare_jump(Op::JumpFalsePop(i16::MAX));

        // <consequent>
        self.block(&if_.consequent)?;

        let Some(alt) = &if_.alternative else {
            // 'skip_conseq:
            self.chunk.set_jump_target(skip_conseq);
            self.chunk.push(Op::Nil);
            return Ok(());
        };

        // jump 'skip_alt
        let skip_alt = self.chunk.prepare_jump(Op::Jump(i16::MAX));

        // 'skip_conseq:
        self.chunk.set_jump_target(skip_conseq);

        // <alternative>
        self.block(alt)?;

        // 'skip_alt:
        self.chunk.set_jump_target(skip_alt);

        Ok(())
    }

    #[instrument(level = "trace")]
    fn identifier(&mut self, ident: &Identifier) -> Fallible<()> {
        if let Some((slot, _local)) = self.locals.resolve(&ident.name) {
            self.chunk.push(Op::LocalGet(slot));
        } else {
            let global = self.chunk.make_global(ident.name.clone());
            self.chunk.push(Op::GlobalGet(global));
        }
        Ok(())
    }

    #[instrument(level = "trace")]
    fn literal(&mut self, lit: &Literal) -> Fallible<()> {
        match lit {
            Literal::Nil => {
                self.chunk.push(Op::Nil);
            }
            Literal::Boolean(b) => {
                if *b {
                    self.chunk.push(Op::True);
                } else {
                    self.chunk.push(Op::False);
                }
            }

            Literal::Integer(v) => {
                let r = BigRational::new(v.clone(), 1.into());
                let id = self.chunk.add_constant(Constant::Rational(r));
                self.chunk.push(Op::Constant(id));
            }
            Literal::Float(v) => {
                let id = self.chunk.add_constant(Constant::Float(*v));
                self.chunk.push(Op::Constant(id));
            }
            Literal::String(v) => {
                let id = self.chunk.add_constant(Constant::String(v.clone()));
                self.chunk.push(Op::Constant(id));
            }
        }

        Ok(())
    }
}

#[derive(Debug, Clone)]
struct Local {
    depth: usize,
    initialized: bool,

    ident: Identifier,
}

#[derive(Debug, Clone, Default)]
struct Locals {
    locals: Vec<Local>,
    scope_depth: usize,
}

impl Locals {
    #[instrument(level = "trace")]
    fn begin_scope(&mut self) {
        self.scope_depth += 1;
    }

    #[instrument(level = "trace")]
    fn end_scope(&mut self) -> usize {
        self.scope_depth -= 1;

        let mut count = 0;

        while let Some(local) = self.locals.last() {
            if local.depth > self.scope_depth {
                self.locals.pop();
                count += 1;
            } else {
                break;
            }
        }

        count
    }

    fn iter_slots(&self) -> impl Iterator<Item = (usize, &Local)> {
        self.locals.iter().enumerate().rev()
    }

    fn iter_stack(&self) -> impl Iterator<Item = &Local> {
        self.locals.iter().rev()
    }

    #[instrument(level = "trace")]
    fn resolve<'a>(&'a self, name: &str) -> Option<(u8, &'a Local)> {
        for (slot, var) in self.iter_slots() {
            if var.ident.name == name {
                let slot: u8 = slot.try_into().expect("Too many locals!");
                return Some((slot, var));
            }
        }
        None
    }

    #[instrument(level = "trace")]
    fn reserve(&mut self, ident: Identifier) -> CompileResult<u8> {
        if self.locals.len() >= (u8::MAX as usize) {
            panic!("Too many locals! Time to add Local2 variants");
        }

        let new = Local {
            ident,
            depth: self.scope_depth,
            initialized: false,
        };

        for old in self.iter_stack() {
            if old.ident.name == new.ident.name && old.depth == new.depth {
                return Err(CompileError::DuplicateVariable(new.ident));
            }
        }

        let slot: u8 = self
            .locals
            .len()
            .try_into()
            .expect("Too many locals! Time to add Local2 variants");
        self.locals.push(new);
        Ok(slot)
    }

    #[instrument(level = "trace")]
    fn mark_init(&mut self, slot: u8) {
        self.locals[slot as usize].initialized = true;
    }
}
