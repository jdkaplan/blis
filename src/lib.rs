pub mod bytecode;
pub mod compile;
pub mod parse;
pub mod runtime;
pub mod vm;

#[cfg(test)]
mod testing;

#[cfg(test)]
mod tests {
    use std::path::PathBuf;

    use rstest::rstest;

    use crate::testing::{replace_object_ids, snapshot_name};

    use super::compile::Compiler;
    use super::parse::Parser;
    use super::vm::{Vm, VmOptions, VmResult};

    struct TestResult {
        result: VmResult<()>,
        stdout: String,
        stderr: String,
    }

    fn run(source: &str) -> TestResult {
        let ast = Parser::parse(source).unwrap();
        let program = Compiler::compile(&ast).unwrap();

        let mut stdout = Vec::new();
        let mut stderr = Vec::new();

        let result = {
            let mut vm = Vm::new(VmOptions {
                stdout: Box::new(&mut stdout),
                stderr: Box::new(&mut stderr),
            });
            vm.interpret(program)
        };

        TestResult {
            result,
            stdout: replace_object_ids(String::from_utf8(stdout).unwrap()),
            stderr: replace_object_ids(String::from_utf8(stderr).unwrap()),
        }
    }

    #[rstest]
    fn test_programs(#[files("test_programs/*.blis")] path: PathBuf) {
        let source = std::fs::read_to_string(&path).unwrap();
        let run = run(&source);

        assert!(run.result.is_ok(), "{:?}", run.stderr);

        insta::with_settings!({
            input_file => &path,
            description => source,
            omit_expression => true,
        }, {
            insta::assert_snapshot!(snapshot_name(&path, "stdout"), run.stdout);
        })
    }

    #[rstest]
    fn test_runtime_errors(#[files("test_programs/err_runtime/*.blis")] path: PathBuf) {
        let source = std::fs::read_to_string(&path).unwrap();
        let run = run(&source);

        assert!(run.result.is_err());

        insta::with_settings!({
            input_file => &path,
            description => source,
            omit_expression => true,
        }, {
            insta::assert_snapshot!(snapshot_name(&path, "error"), run.result.unwrap_err());
            insta::assert_snapshot!(snapshot_name(&path, "stdout"), run.stdout);
            insta::assert_snapshot!(snapshot_name(&path, "stderr"), run.stderr);
        })
    }
}
