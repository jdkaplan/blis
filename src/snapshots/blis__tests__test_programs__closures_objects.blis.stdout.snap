---
source: src/lib.rs
description: "func main() {\n    let one = new_thing(\"uno\");\n    let two = new_thing(\"dos\");\n\n    println(one);\n    println(one.id, one.name);\n\n    println(two);\n    println(two.id, two.name);\n\n    one.two = two;\n    two.one = one;\n\n    println(one.two);\n    println(two.one);\n}\n\nlet counter = 0;\n\nfunc new_thing(name) {\n    counter = counter + 1;\n\n    object {\n        id = counter,\n        name = name,\n    }\n}\n\nmain();\n"
input_file: test_programs/closures_objects.blis
---
<instance of object 0x000000000001>
1 uno
<instance of object 0x000000000002>
2 dos
<instance of object 0x000000000002>
<instance of object 0x000000000001>
